File: example-java-android/src/main/java/com/vimeo/android/networking/example/AccountPreferenceManager.java
package com.vimeo.android.networking.example;

import android.content.Context;
import android.content.SharedPreferences;
import android.preference.PreferenceManager;
import android.support.annotation.Nullable;

import com.vimeo.networking.utils.VimeoNetworkUtil;
import com.vimeo.networking2.User;
import com.vimeo.networking2.VimeoAccount;

import java.io.IOException;

/**
 * Class for persisting the the auth access token (and possibly the user)
 * <p/>
 * Created by kylevenn on 5/26/15.
 */
public class AccountPreferenceManager {

    private static SharedPreferences sSharedPreferences;

    public static synchronized void initializeInstance(Context context) {
        if (sSharedPreferences == null) {
            sSharedPreferences = PreferenceManager.getDefaultSharedPreferences(context);
        }
    }

    // <editor-fold desc="Account">

    public static final String CLIENT_ACCOUNT_JSON = "CLIENT_ACCOUNT_JSON";
    public static final String CACHED_CLIENT_CREDENTIALS_ACCOUNT_JSON =
            "CACHED_CLIENT_CREDENTIALS_ACCOUNT_JSON";

    public static void removeClientAccount() {
        sSharedPreferences.edit().remove(CLIENT_ACCOUNT_JSON).apply();
    }

    @Nullable
    public static VimeoAccount getClientAccount() throws IOException {
        // NOTE: This happens on the main thread, don't do this
        String accountJSON = sSharedPreferences.getString(CLIENT_ACCOUNT_JSON, null);
        return accountJSON == null ? null : VimeoNetworkUtil.getMoshi()
                .adapter(VimeoAccount.class).fromJson(accountJSON);
    }

    public static void setClientAccount(VimeoAccount vimeoAccount) {
        if (vimeoAccount != null) {
            // NOTE: This happens on the main thread, don't do this
            String accountJSON = VimeoNetworkUtil.getMoshi()
                    .adapter(VimeoAccount.class).toJson(vimeoAccount);
            sSharedPreferences.edit().putString(CLIENT_ACCOUNT_JSON, accountJSON).apply();
        }
    }

    @Nullable
    public static User getCurrentUser() throws IOException {
        VimeoAccount clientAccount = AccountPreferenceManager.getClientAccount();
        return clientAccount != null ? clientAccount.getUser() : null;
    }

    public static void cacheClientCredentialsAccount(VimeoAccount vimeoAccount) {
        if (vimeoAccount != null) {
            String accountJSON = VimeoNetworkUtil.getMoshi()
                    .adapter(VimeoAccount.class).toJson(vimeoAccount);
            sSharedPreferences.edit().putString(CACHED_CLIENT_CREDENTIALS_ACCOUNT_JSON, accountJSON).apply();
        }
    }

    public static VimeoAccount getCachedClientCredentialsAccount() throws IOException {
        String accountJSON = sSharedPreferences.getString(CACHED_CLIENT_CREDENTIALS_ACCOUNT_JSON, null);
        return accountJSON == null ? null : VimeoNetworkUtil.getMoshi()
                .adapter(VimeoAccount.class).fromJson(accountJSON);
    }

    // </editor-fold>

}

File: example-kotlin-android/src/main/kotlin/com/vimeo/android/networking/example/kotlin/AccountPreferenceManager.kt
package com.vimeo.android.networking.example.kotlin

import android.content.Context
import android.content.SharedPreferences
import android.preference.PreferenceManager
import com.vimeo.networking.utils.VimeoNetworkUtil
import com.vimeo.networking2.User
import com.vimeo.networking2.VimeoAccount

/**
 * Class for persisting the the auth access token (and possibly the user)
 *
 * Created by anthonyr on 5/8/17.
 */
@Suppress("unused")
object AccountPreferenceManager {

    private lateinit var sharedPreferences: SharedPreferences

    @Synchronized
    fun initializeInstance(context: Context) {
        sharedPreferences = PreferenceManager.getDefaultSharedPreferences(context)
    }

    // <editor-fold desc="Account">

    private const val CLIENT_ACCOUNT_JSON = "CLIENT_ACCOUNT_JSON"
    private const val CACHED_CLIENT_CREDENTIALS_ACCOUNT_JSON = "CACHED_CLIENT_CREDENTIALS_ACCOUNT_JSON"

    fun removeClientAccount() {
        sharedPreferences.edit().remove(CLIENT_ACCOUNT_JSON).apply()
    }

    // NOTE: This happens on the main thread, don't do this
    // NOTE: This happens on the main thread, don't do this
    var clientAccount: VimeoAccount?
        get() {
            val accountJSON = sharedPreferences.getString(CLIENT_ACCOUNT_JSON, null) ?: return null

            return VimeoNetworkUtil.getMoshi()
                    .adapter(VimeoAccount::class.java).fromJson(accountJSON)
        }
        set(vimeoAccount) {
            if (vimeoAccount != null) {
                VimeoNetworkUtil.getMoshi()
                        .adapter(VimeoAccount::class.java).toJson(vimeoAccount)?.also {
                    sharedPreferences.edit().putString(CLIENT_ACCOUNT_JSON, it).apply()
                } ?: removeClientAccount()
            }
        }

    val currentUser: User? get() = clientAccount?.user

    fun cacheClientCredentialsAccount(vimeoAccount: VimeoAccount?) {
        vimeoAccount?.run {
            val accountJSON = VimeoNetworkUtil.getMoshi()
                    .adapter(VimeoAccount::class.java).toJson(vimeoAccount) ?: return
            sharedPreferences.edit().putString(CACHED_CLIENT_CREDENTIALS_ACCOUNT_JSON, accountJSON).apply()
        }
    }

    val cachedClientCredentialsAccount: VimeoAccount?
        get() {
            val accountJSON = sharedPreferences.getString(CACHED_CLIENT_CREDENTIALS_ACCOUNT_JSON, null)
                ?: return null

            return VimeoNetworkUtil.getMoshi()
                    .adapter(VimeoAccount::class.java).fromJson(accountJSON)
        }

    // </editor-fold>

}
